lib/dry/schema/primitive_inferrer.rb

```diff
@@ -0,0 +1,98 @@
+# frozen_string_literal: true
+
+require 'dry/core/cache'
+
+module Dry
+  module Schema
+    # PrimitiveInferrer is used internally by `Macros::Filled`
+    # for inferring a list of possible primitives that a given
+    # type can handle.
+    #
+    # @api private
+    class PrimitiveInferrer
+      extend Dry::Core::Cache
+
+      # Compiler reduces type AST into a list of primitives
+      #
+      # @api private
+      class Compiler
+        # @api private
+        def visit(node)
+          meth, rest = node
+          public_send(:"visit_#{meth}", rest)
+        end
+
+        # @api private
+        def visit_nominal(node)
+          type, _ = node
+          [type]
+        end
+
+        # @api private
+        def visit_hash(_)
+          Hash
+        end
+
+        # @api private
+        def visit_array(_)
+          Array
+        end
+
+        # @api private
+        def visit_lax(node)
+          visit(node)
+        end
+
+        # @api private
+        def visit_constructor(node)
+          other, * = node
+          visit(other)
+        end
+
+        # @api private
+        def visit_enum(node)
+          other, * = node
+          visit(other)
+        end
+
+        # @api private
+        def visit_sum(node)
+          left, right = node
+
+          [visit(left), visit(right)].flatten
```

Comment:
I would write `flatten(1)` because it should suffice here. If it's not working there must be something wrong

Reply:
I think it can be deeply nested when there are deeply nested sums like `Integer | String | Array`, no?

Reply:
yes but every level has to be balanced with one `visit_sum`, as in `Integer | String | Array` will make two `visit_sum` calls (by the number of `|`)